graph.go

```diff
@@ -12,7 +12,7 @@ import (
 type graph struct {
 
 	// roots maps PipelineIDs to root Nodes
-	roots map[PipelineID]*linkedNode
+	roots sync.Map // PipelineID -> *linkedNode
```

Comment:
It would be pretty trivial to wrap the sync.Map type inside a type-safe version with the same interface. Obviously `roots` is not an exported field, so the risk should be pretty negligible, but given the wide usage of this library, I was thinking about whether it might be worth considering the extra type safety for robustness? I'm happy to hear pushback too though.

Reply:
Done.